crates/matrix-sdk/src/sliding_sync/list/mod.rs

```diff
@@ -929,12 +710,40 @@ pub enum SlidingSyncState {
     FullyLoaded,
 }
 
+/// Builder for a new sliding sync list in selective mode.
+///
+/// Conveniently allows to add ranges.
+#[derive(Clone, Debug, Default)]
+pub struct SlidingSyncSelectiveModeBuilder {
```

Comment:
We are facing 2 patterns now.

1. Either we use a builder pattern to configure the sync-mode (with `SlidingSyncSelectiveModeBuilder` as we have now),
2. Or, we use a constructor (with `SlidingSyncMode::new_selective`, `::new_growing` and `::new_paging`).

Note that `new_growing` and `new_paging` are using arguments to configure the few options (`batch_size` and `maximum_number_of_rooms_to_fetch`). But, in a very close future, I want to introduce `skip`/`offset`, which will be very useful when we jump from Selective mode to Growing mode for example (e.g. we start with a range of `0..=19` in Selective mode, then switch to Growing mode with `batch_size = 50`, it will be re-request the first 20 items, hence `skip = 20`, tadaa).

So either we keep going with the constructors, in this case:

* `::new_selective(ranges: Vec<RangeInclude<…>>`
* `::new_growing(batch_size, maximum_number_of_rooms_to_fetch, /*later*/ skip)`
* `::new_paging(batch_size, maximum_number_of_rooms_to_fetch, /*later*/ skip)`

Or, we use builders. I like having builders because we can have default values and makes the code cleaner. But it's also _more code_.

Alternatively, we can use the constructors to… construct a builder, something like:

```rust
list.set_sync_mode(SlidingSyncMode::new_selective().add_range(0..=19));
list.set_sync_mode(SlidingSyncMode::new_growing().batch_size(50).skip(20);
```

We can finish with a `.build()`, or just implement a `From<SlidingSyncModeSelectiveBuilder> for SlidingSyncMode`, and get `fn set_sync_mode<B>(sync_mode_builder: B) where B: Into<SlidingSyncMode>`. That would be quite generic.

What do you think?

Reply:
Yep, it's slightly nicer indeed. I think adding the builder pattern could be done in the future for the paginated/growing modes, since right now it would be indeed a bit verbose only for the two parameters those have.

Reply:
Right now, for `Growing` and `Paging`, we are always specifying the `maximum_number_of_rooms_to_fetch`, and it's most of the time `None`. I can write the builders for those in a different PR, I'm fine with that.

Reply:
I'll add that quickly now!